src/gz/mod.rs

```diff
@@ -508,6 +508,85 @@ mod tests {
         assert_eq!(res, v);
     }
 
+    // A Rust implementation of CRC that closely matches the C code in RFC1952.
+    // Only use this to create CRCs for tests.
+    struct Rfc1952Crc {
+        /* Table of CRCs of all 8-bit messages. */
+        crc_table: [u32; 256],
+    }
+
+    impl Rfc1952Crc {
+        fn new() -> Self {
+            let mut crc = Rfc1952Crc {
+                crc_table: [0; 256],
+            };
+            /* Make the table for a fast CRC. */
+            for n in 0usize..256 {
+                let mut c = n as u32;
+                for _k in 0..8 {
+                    if c & 1 != 0 {
+                        c = 0xedb88320 ^ (c >> 1);
+                    } else {
+                        c = c >> 1;
+                    }
+                }
+                crc.crc_table[n] = c;
+            }
+            crc
+        }
+
+        /*
+         Update a running crc with the bytes buf and return
+         the updated crc. The crc should be initialized to zero. Pre- and
+         post-conditioning (one's complement) is performed within this
+         function so it shouldn't be done by the caller.
+        */
+        fn update_crc(&self, crc: u32, buf: &[u8]) -> u32 {
+            let mut c = crc ^ 0xffffffff;
+
+            for b in buf {
+                c = self.crc_table[(c as u8 ^ *b) as usize] ^ (c >> 8);
+            }
+            c ^ 0xffffffff
+        }
+
+        /* Return the CRC of the bytes buf. */
+        fn crc(&self, buf: &[u8]) -> u32 {
+            self.update_crc(0, buf)
+        }
+    }
+
+    #[test]
+    fn roundtrip_header() {
+        let mut header = GzBuilder::new()
+            .mtime(1234)
+            .operating_system(57)
+            .filename("filename")
+            .comment("comment")
+            .into_header(Compression::fast());
+
+        // Add a CRC to the header
+        header[3] = header[3] ^ super::FHCRC;
```

Comment:
I wonder if this header-CRC feature is something that should be natively supported one day. Right now I am sure (nearly) nobody would ever do this, even though it seems like a feature people might find interesting to harden their streams.

Maybe that's an easy win for future PRs.

Reply:
While it would be a simple change, there doesn't seem to be demand for it. Neither the `gzip` command nor the Python `gzip` library create a header CRC. Which explains why this bug hasn't caused any problems for real gzipped files.